security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -307,13 +231,252 @@ func TestAESGCMInvalidKeySize(t *testing.T) {
 	}
 }
 
-func dehex(s string) []byte {
-	if len(s) == 0 {
-		return make([]byte, 0)
+// Test encrypt and decrypt on roundtrip messages for AES-GCM using an updated
+// key.
+func TestAESGCMUpdatedKey(t *testing.T) {
+	for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
+		key := make([]byte, keySize)
+		client, server := getGCMCryptoPair(key, t)
+		// Update the key with a new one which is different from the original.
+		newKey := make([]byte, keySize)
+		newKey[0] = '\xbd'
+		err := client.UpdateKey(newKey)
+		if err != nil {
+			t.Fatalf("client UpdateKey failed with: %v", err)
+		}
+		err = server.UpdateKey(newKey)
+		if err != nil {
+			t.Fatalf("server UpdateKey failed with: %v", err)
+		}
+		testGCMEncryptRoundtrip(client, server, t)
+	}
+}
+
+// Test update key for AES-GCM using a key with different size from the initial
+// key.
+func TestAESGCMMismatchKeySizeUpdate(t *testing.T) {
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(key) = %v", err)
+	}
+
+	// Update the key with a new one which is a different from the original.
+	newKey := make([]byte, aes256GcmKeySize)
+	err = crypter.UpdateKey(newKey)
+	if err == nil {
+		t.Fatal("UpdateKey should fail with invalid key size error")
 	}
-	b, err := hex.DecodeString(s)
+}
+
+// Test update key for AES-GCM using a key with an invalid size.
+func TestAESGCMInvalidKeySizeUpdate(t *testing.T) {
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
 	if err != nil {
-		panic(err)
+		t.Fatalf("NewAESGCM(key) = %v", err)
+	}
+
+	// Update the key with a new one which is an invalid size.
+	newKey := make([]byte, 17)
+	err = crypter.UpdateKey(newKey)
+	if err == nil {
+		t.Fatal("UpdateKey should fail with invalid key size error")
+	}
+}
+
+// Test update key for AES-GCM using a key with an invalid size.
+func TestAESGCMNISTAndIEEE(t *testing.T) {
+	// NIST vectors from:
+	// http://csrc.nist.gov/groups/ST/toolkit/BCM/documents/proposedmodes/gcm/gcm-revised-spec.pdf
+	// IEEE vectors from:
+	// http://www.ieee802.org/1/files/public/docs2011/bn-randall-test-vectors-0511-v1.pdf
+	for _, test := range []CryptoTestVector{
+		{
+			Comment:    "Derived from NIST test vector 1",
```

Comment:
Aren't these the actual NIST/IEEE test vectors, and not just a test vector derived from them?

Reply:
Yes I believe so. The comment was from the Java/C++ NIST/IEEE unit tests which I converted to Go. I left the comment as it originally was in the Java version. I'll remove the `Derived from`.

Reply:
SGTM. Thanks!